### ISSUE TITLE
Sign up flow changes: User’s Country (of residence)

### ISSUE THREAD

**Issue by pe0ny** [toggl/mobileapp#1610]
Regarding GDPR (General Data Protection Regulation) changes we will need 2 additions to current sign up flow: User’s Country of residence & active consent to our Terms of Service.

## User’s Country (of residence):
- User can access list of countries when tapping the Country dropdown at the top of the screen on the Sign up page #1960
- List of countries comes from BE. They will modify the v9/signups endpoint (as part of https://github.com/toggl/backend/issues/6354) to include this Country information. (It has country codes to match geolocation against:  https://github.com/toggl/backend/issues/6348)
- Always pre-select the country into the form, based on location info we can get.
- A country can be changed but not removed (no option to choose “No country”, “Blank”, “Other” or similar).
- Allow user to search countries by typing.
- User can navigate back to the Sign up page by:
   - Tapping the back arrow at the top left corner
   - Selecting another country
- If for some reason, we can’t get a location, display text “Select country...” in the dropdown on the Sign up page #1960
- ~~Signup **can be** posted without the country, though including a country is preferred.~~ **EDIT:** posting the country will be mandatory in the future, so we cannot allow the user to go through without having one selected
- For Google Signup, post the preselected (if we can get a location) country along with Signup information (no need to ask for anything extra from Google).
- Display current selection as the first item in list.

### Country of Residence view
![toggl - ios - countries 1x](https://user-images.githubusercontent.com/5064445/39252849-3656d822-48af-11e8-9926-1e07ce7858c2.png)

Zeplin: https://zpl.io/2y76XO9

**Comment by amulware** [comment]
How do we determine the selected country? By IP (how?)? By location (need to ask for permission - very intrusive)? The locale the user selected in device settings (does not require permissions)? Other?

**Comment by amulware** [comment]
Answering myself: https://toggl.com/api/v9/me/location

**Comment by amulware** [comment]
Also, for the list of countries: https://toggl.com/api/v9/countries

**Comment by amulware** [comment]
Match countries by country code.

**Comment by pe0ny** [comment]
> what if the user selects a different country and then taps ‘sign up with google’? will we send THAT country, instead of the preselected one?

post the country the user selected if they did it manually

> I take it the TOS/PP popup here https://github.com/toggl/mobileapp/issues/1611 is shown before we send anything to backend, and we only actually sign up if the user clicks ‘I agree’?

yes

> How does this work with google signup? Do we show the same popup before we _actually_ take them to google?

it will be: tos->google->google signup

**Comment by pe0ny** [comment]
Update: Selected item should be marked as so and displayed as the first item in list:
e.g. like in the Tags view

![toggl - ios - countries 1x](https://user-images.githubusercontent.com/5064445/39696027-19e2feec-51f5-11e8-90dc-13bad89051c6.png)

**Comment by pe0ny** [comment]
> What if fetching countries fails?

Then we won't allow user to sign up and show a default message: "Sorry, we’re having trouble signing you up. Please try again.
At the same time we should try to fetch again (in every 2sec or so), though it's rather likely that if the country fetch fails the signup post will fail as well.

> What if the user selects a country faster than the pre-select-fetch works?

Then we'll keep what user selected and won't override it.

> Does the signup ui wait until we get the list of countries and the user’s preselected one before the user can start interacting with it? if it waits, how do we show this?

"Sign up for free" button should be inactive when the country isn't selected. So while we're fetching countries, user can't interact with it. At the same time, at the top of the screen we should display "Select country..." option. When user taps on it, we can open the Country view (https://zpl.io/2y76XO9) and show a loading state there.
Or another option: When user tries to tap "Sign up for free" without the country being selected, we can show a tooltip which pointing to select a country or little exclamation mark next to the country selection. And when user opens it, we show a loading state if the fetching indeed takes a lot of time.

These are my ideas.
@fonorobert feel free to jump in.

**Comment by amulware** [comment]
Scratch all that. After more discussion we'll go with:
```
1. we hardcode the countries (in the future we can automate fetching these every build)
2. we now can always show the countries list even when offline
3. we try and fetch the user’s location with the other endpoint to preselect from that list - we do this only once when the user comes to the signup screen. if it fails too bad, the user has to select themselves
3.5 when/if we hear back from that endpoint, we select that country from the list, if no country is selected yet (i.e. by the user manually) - in most cases that'll be within a split second
4. user can always select a different country anyways
5. we block ‘sign up’ buttons until a country is selected (automatically, or by the user)
```

**Comment by pe0ny** [comment]
Just to be sure: We'll block both sign up buttons ("Sign up for free" and "Sign up with Google") until a country is selected.

**Comment by amulware** [comment]
Something to consider: if the user still taps the button, and _any_ of the data isn't filled in, it should maybe highlight the missing field or country selection.

Not needed for the MVP, but if you like this, please create an issue and animation @pe0ny :)
